### ISSUE TITLE
CPU bottleneck  

### ISSUE THREAD

**Issue by dmitrykudinov** [fizyr/keras-retinanet#249]
I think there is a bottleneck in Python processing - most of the time it's doing something in a single thread consuming 100% of a single CPU core. Then, briefly, it gives a job to the GPU... but with a powerful GPU, most of the time is spent in CPU, causing a major slowdown:

![image](https://user-images.githubusercontent.com/3879467/35480112-873f6386-03bc-11e8-9359-907c9cbc92eb.png)

This becomes particularly visible with backbone frozen, as there is much fewer parameters to train, so, unfortunately, it pretty much cancels the performance advantage of Transfer Learning...

I think if you could minimize the CPU processing, the module would be an order of magnitude faster on training when working with good GPUs.

**Comment by de-vri-es** [comment]
Thanks for sharing, this is quite interesting. Did you play with any of the multithreading/processing parameters of keras? I don't think can help much for various reasons, but it's good to know anyway.

A good guess for where most of the CPU time is spent would be data loading and preprocessing. However, it seems a bit excessive that it would take multiple seconds to load and preprocess a single batch. What batch size did you train with?

One other possible problem is that you might be using a backend that actually doesn't use your GPU. Maybe you've got the CPU version of tensorflow installed.

If the overhead is in the data preprocessing, I think there are two ways to try and improve it:

* Try to exploit multiple cores. Currently this is quite difficult thanks to the Python global interpreter lock and a general inability to concurrently run the same generator in Python. Multiprocessing would require passing all relevant data over pipes between processes, which can cause significant overhead.

* Try to reduce CPU usage either by being smart about things or by trying to offload more work to the GPU when possible.

**Comment by dmitrykudinov** [comment]
@de-vri-es thanks for the suggestions.

>What batch size did you train with?

from 1 to 48 

> you might be using a backend that actually doesn't use your GPU

I am using this machine with other core Keras / pure TF projects and those always load the GPU 95-100%, sometimes for weeks.... so I do not think the problem is in my configuration.

if you can disclose, what GPU do you use to train this implementation?

**Comment by dmitrykudinov** [comment]
IMHO, using generators is a bad idea, not suitable for production-grade systems. If I need data augmentation, this needs to be done as a separate preprocessing step, using different means than just a  single-threaded python script. And then, once I got the preprocessing done, I can just feed straight the batches of preliminary augmented images into the training pipeline, focusing on tuning the NN architecture, hyperparams, instead of being distracted by slow on-the-fly augmentation... Just my 2c.

**Comment by de-vri-es** [comment]
> IMHO, using generators is a bad idea, not suitable for production-grade systems. If I need data augmentation, this needs to be done as a separate preprocessing step, using different means than just a single-threaded python script. And then, once I got the preprocessing done, I can just feed straight the batches of preliminary augmented images into the training pipeline, focusing on tuning the NN architecture, hyperparams, instead of being distracted by slow on-the-fly augmentation... Just my 2c.

Well, the augmentation is optional, although not currently exposed that way through the train script. Could you test without any augmentation by turning this line into `transform_generator = None`?
https://github.com/fizyr/keras-retinanet/blob/b6b1b593eef8ad85c2f18ccb3213316fc8a2dfaf/keras_retinanet/bin/train.py#L119

**Comment by lvaleriu** [comment]
An idea would be to pack images into record files (i have in mind the mxnet record_io files). Would you like to see such a pull request? (Adding a new generator - like the csv generator - that loads the data from a single record file).
Mxnet have also Image iterators with a large number of augmentation choices for detection (written in C++ and with multiple workers option - and they are FAST).

Here is an interesting chapter called "Designing Efficient Data Loaders for Deep Learning" https://mxnet.incubator.apache.org/architecture/note_data_loading.html that explains the optimisations of such process.

I'm actually doing this on my own since i'm trying to train retinanet on a bigger scale dataset: OpenImagesDataset. (Btw, I've just added a new pullrequest for this dataset)

**Comment by lvaleriu** [comment]
The record files is useful if the dataset is located on a HDD. If you have a fast SSD or even a SSD NVME M2 disk this doesnt bring much gain in speed.

I have a Samsung Pro SSD disk and with a batch size of 4 the training on OpenImages dataset (1200x800 like images) uses 99%, 100% of a PASCAL TITAN Gpu (continously).

But if you want to do combine many image augmentations it is still better to have a tool like the mxnet augmenters and of course a good multi-core cpu :)

**Comment by de-vri-es** [comment]
Thanks for the suggestions. Before choosing a specific direction I would like to see/do some benchmarks to identify the bottleneck[s] we currently have. Data augmentation is a prime suspect, but I'd like to be sure.

We're also training mainly using Titan X cards and GTX 1080's or something along those lines. With those cards we can keep them near 100% continuously, although I'm not sure if that is still true with higher batch sizes and the backbone layers frozen. Maybe @hgaiser knows.

> An idea would be to pack images into record files (i have in mind the mxnet record_io files). Would you like to see such a pull request? (Adding a new generator - like the csv generator - that loads the data from a single record file).
> Mxnet have also Image iterators with a large number of augmentation choices for detection (written in C++ and with multiple workers option - and they are FAST).

If there is a clear advantage to using mxnet record files, it might be an idea. However, I'd also like to keep the number of dependencies for `keras-retinanet` small. In an ideal world we wouldn't even have to supply data generators and training scripts, but focus purely on the network architecture.

**Comment by dmitrykudinov** [comment]
@de-vri-es yes, disabling generator per your suggestion brings the GPU (I am using NVidia GP100) usage to consistent 98-100% , reducing the training time per epoch from about 40 minutes to 11. So, i would call it confirmed. Thanks for the suggestion.

**Comment by dmitrykudinov** [comment]
@de-vri-es verified, https://github.com/fizyr/keras-retinanet/pull/246 fixed the problem. closing... thanks!

**Comment by de-vri-es** [comment]
Good to hear. Thanks for testing!

**Comment by MarkAlive** [comment]
> offload more work to the GPU when possible.
@de-vri-es Do you have any demos about it? Thanks a lot for sharing.

**Comment by anil-bit** [comment]
dmitrykudinov , did you find the solution?
